tpu/templates/tpu_film_estimator/trainer.py

```diff
@@ -0,0 +1,266 @@
+    conv_0 = tf.layers.conv2d(x, filters=filter_0, kernel_size=(3, 3))
+
+    # Apply FiLM before the ReLU activation.
+    # Reshape the modulation parameters manually.
+    gamma_0 = all_gamma[:, None, None, :filter_0]
+    beta_0 = all_beta[:, None, None, :filter_0]
+    filmed_conv_0 = FeaturewiseLinearModulationLayer()(conv_0, gamma_0, beta_0)
+
+    conv_out_0 = tf.nn.relu(filmed_conv_0)
+
+    # Do the same for the next convolutional block
+    filter_1 = filter_sizes[1]
+    conv_1 = tf.layers.conv2d(conv_out_0, filters=filter_1, kernel_size=(3, 3))
+
+    gamma_1 = all_gamma[:, None, None, -filter_1:]
+    beta_1 = all_beta[:, None, None, -filter_1:]
+    filmed_conv_1 = FeaturewiseLinearModulationLayer()(conv_1, gamma_1, beta_1)
+
+    conv_out_1 = tf.nn.relu(filmed_conv_1)
+
+    # Fully connected logits output
+    flattened = tf.reshape(conv_out_1, (params['batch_size'], -1))
+    label_classification_logits = tf.layers.dense(flattened, n_classes)
+
+    predictions = tf.nn.softmax(label_classification_logits)
+    loss = None
+    train_op = None
+
+    if mode == tf.estimator.ModeKeys.TRAIN:
+        # define loss
+        loss = tf.losses.softmax_cross_entropy(
+            onehot_labels=onehot_labels,
+            logits=label_classification_logits
+        )
+
+        # define train_op
+        optimizer = tf.train.RMSPropOptimizer(learning_rate=0.05)
+
+        # wrapper to make the optimizer work with TPUs
+        if params['use_tpu']:
+            optimizer = tf.contrib.tpu.CrossShardOptimizer(optimizer)
+
+        train_op = optimizer.minimize(loss, global_step=global_step)
+
+    if params['use_tpu']:
+        # TPU version of EstimatorSpec
+        return tf.contrib.tpu.TPUEstimatorSpec(
+            mode=mode,
+            predictions=predictions,
+            loss=loss,
+            train_op=train_op)
+    else:
+        return tf.estimator.EstimatorSpec(
+            mode=mode,
+            predictions=predictions,
+            loss=loss,
+            train_op=train_op)
+
+
+# ## The input function
+#
+
+def train_input_fn(params={}):
+    # labaled image data
+    x = np.random.rand(100, 28, 28, 3)
+    y = np.random.randint(0, n_classes, 100)
+
+    # additional input data for modulation
+    modulation_data = np.random.rand(100, 5)
+
+    x_tensor = tf.constant(x, dtype=tf.float32)
+    y_tensor = tf.constant(y, dtype=tf.int32)
+    modulation_data_tensor = tf.constant(modulation_data, dtype=tf.float32)
+
+    # make a dataset
+    dataset = tf.data.Dataset.from_tensor_slices((x_tensor, y_tensor, modulation_data_tensor))
+
+    # TPUEstimator passes params when calling input_fn
+    batch_size = params.get('batch_size', 16)
+
+    dataset = dataset.repeat().shuffle(32).batch(batch_size, drop_remainder=True)
+
+    # TPUs need to know all dimensions when the graph is built
+    # Datasets know the batch size only when the graph is run
+    def set_shapes_and_format(x, y, modulation_data):
```

Comment:
Can we add Documentation information to the function as:
```
def foo():
  """This is a foo function which does x.

   Args:
      x: (type) This an argument.

   Returns:
      A foo value.
  """
```

Reply:
this is an inner function that probably would have been a `lambda`, as opposed to a library function to be reused out of context.  i feel the preceeding comment clarifies its purpose well enough.  i added a docstring.

Reply:
I agree, comment is self-explanatory. Just would like to standardize our samples and use a common structure.